File: C++/2-Medium/00275._H_指数_II/00275._H_指数_II.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int hIndex(vector<int>& citations) {
        int left = 0, right = citations.size() - 1, mid;
        while (left <= right) {
            mid = left + (right - left) / 2;
            // 如果这篇论文的引用次数小于等于其后面论文的数量，说明H指数至少为citations.size() - mid
            if (citations[mid] < citations.size() - mid) left = mid + 1;
            else right = mid - 1;
        }
        // 数组长度减去左边界位置，得到H指数
        return citations.size() - left;
    }
};

int main() {
    // 示例：引用次数数组
    vector<int> citations = {0, 1, 3, 5, 6};
    Solution solution;
    // 计算并输出H指数
    cout << "H-Index is: " << solution.hIndex(citations) << endl;
    return 0;
}

File: Python/2-Medium/00275._H_指数_II/00275._H_指数_II.py
from typing import List

def h_index(citations: List[int]) -> int:
    """
    使用二分查找算法计算 H 指数。

    Args:
    citations (List[int]): 已排序的引用次数数组。

    Returns:
    int: 研究者的 H 指数。
    """
    left, right = 0, len(citations) - 1
    while left <= right:
        mid = (left + right) // 2
        # 检查是否满足 H 指数的条件
        if citations[mid] < len(citations) - mid:
            left = mid + 1  # 向右搜索
        else:
            right = mid - 1  # 向左搜索
    # 计算 H 指数
    return len(citations) - left

# 示例运行
citations = [0, 1, 3, 5, 6, 7]
h_index_result = h_index(citations)
print(h_index_result)  # 3